File: src/noise.cpp
#include "noise.hpp"

#include <algorithm>
#include <cmath>
#include <iostream>

namespace volt {

ThermalNoiseInjector::ThermalNoiseInjector(const Config& cfg)
    : cfg_(cfg), rng_(cfg.noise_seed), dist_(0.0f, cfg.noise_stddev) {}

void ThermalNoiseInjector::inject_transient(std::vector<float>& conductances) {
    if (cfg_.noise_stddev <= 0.0f) {
        return;
    }
    for (float& g : conductances) {
        g += dist_(rng_);
    }
}

void ThermalNoiseInjector::inject_persistent(CrossbarArray& array) {
    if (cfg_.noise_stddev <= 0.0f) {
        return;
    }
    const float gmax = array.effective_g_max();
    for (int i = 0; i < array.rows(); ++i) {
        for (int j = 0; j < array.cols(); ++j) {
            float np = dist_(rng_);
            float nn = dist_(rng_);
            array.G_pos_[static_cast<std::size_t>(i)][static_cast<std::size_t>(j)] =
                std::clamp(array.G_pos_[static_cast<std::size_t>(i)][static_cast<std::size_t>(j)] + np,
                           0.0f, gmax);
            array.G_neg_[static_cast<std::size_t>(i)][static_cast<std::size_t>(j)] =
                std::clamp(array.G_neg_[static_cast<std::size_t>(i)][static_cast<std::size_t>(j)] + nn,
                           0.0f, gmax);
        }
    }
}

ReadDisturbSimulator::ReadDisturbSimulator(const Config& cfg) : cfg_(cfg) {}

void ReadDisturbSimulator::apply_disturb(CrossbarArray& array, int active_row, float V_applied) {
    if (active_row < 0 || active_row >= array.rows()) {
        return;
    }
    float V_dis = V_applied * cfg_.disturb_ratio;
    float delta = cfg_.disturb_alpha * V_dis;
    const float gmax = array.effective_g_max();
    for (int neighbor : {active_row - 1, active_row + 1}) {
        if (neighbor < 0 || neighbor >= array.rows()) {
            continue;
        }
        for (int j = 0; j < array.cols(); ++j) {
            array.G_pos_[static_cast<std::size_t>(neighbor)][static_cast<std::size_t>(j)] =
                std::clamp(
                    array.G_pos_[static_cast<std::size_t>(neighbor)][static_cast<std::size_t>(j)] +
                        delta,
                    0.0f, gmax);
            array.G_neg_[static_cast<std::size_t>(neighbor)][static_cast<std::size_t>(j)] =
                std::clamp(
                    array.G_neg_[static_cast<std::size_t>(neighbor)][static_cast<std::size_t>(j)] +
                        delta,
                    0.0f, gmax);
        }
    }
}

void ReadDisturbSimulator::log_drift_report(const CrossbarArray& array) const {
    double sum_shift = 0.0;
    double max_shift = 0.0;
    int n = array.rows() * array.cols();
    if (n == 0) {
        std::cout << "[ReadDisturb] drift report: empty array\n";
        return;
    }
    for (int i = 0; i < array.rows(); ++i) {
        for (int j = 0; j < array.cols(); ++j) {
            float gp0 = array.G_pos_baseline_[static_cast<std::size_t>(i)]
                                      [static_cast<std::size_t>(j)];
            float gn0 = array.G_neg_baseline_[static_cast<std::size_t>(i)]
                                      [static_cast<std::size_t>(j)];
            float gp = array.G_pos_[static_cast<std::size_t>(i)][static_cast<std::size_t>(j)];
            float gn = array.G_neg_[static_cast<std::size_t>(i)][static_cast<std::size_t>(j)];
            double shift_pos = std::abs(static_cast<double>(gp) - static_cast<double>(gp0));
            double shift_neg = std::abs(static_cast<double>(gn) - static_cast<double>(gn0));
            double cell_shift = std::max(shift_pos, shift_neg);
            sum_shift += cell_shift;
            max_shift = std::max(max_shift, cell_shift);
        }
    }
    double avg = sum_shift / static_cast<double>(n);
    std::cout << "[ReadDisturb] drift: avg conductance shift = " << avg
              << " S, max = " << max_shift << " S\n";
}

WriteEnduranceSimulator::WriteEnduranceSimulator(const Config& cfg) : cfg_(cfg) {}

void WriteEnduranceSimulator::apply_write_cycles(CrossbarArray& array, int cycles) {
    if (cycles <= 0 || cfg_.write_endurance_lambda <= 0.0f) {
        return;
    }
    const float scale =
        std::exp(-cfg_.write_endurance_lambda * static_cast<float>(cycles));
    array.apply_uniform_conductance_scale(scale);
}

}  // namespace volt

File: crates/volt-core/src/noise.rs
use rand_distr::{Distribution, Normal};
use rand::SeedableRng;
use rand::rngs::StdRng;

use crate::config::Config;
use crate::crossbar::CrossbarArray;

pub struct ThermalNoiseInjector {
    cfg: Config,
    rng: StdRng,
    dist: Normal<f32>,
}

impl ThermalNoiseInjector {
    pub fn new(cfg: Config) -> Self {
        let dist = Normal::new(0.0, cfg.noise_stddev).expect("invalid noise stddev");
        let rng = StdRng::seed_from_u64(cfg.noise_seed as u64);
        Self { cfg, rng, dist }
    }

    pub fn inject_transient(&mut self, conductances: &mut [f32]) {
        if self.cfg.noise_stddev <= 0.0 {
            return;
        }
        for g in conductances.iter_mut() {
            *g += self.dist.sample(&mut self.rng);
        }
    }

    pub fn inject_persistent(&mut self, array: &mut CrossbarArray) {
        if self.cfg.noise_stddev <= 0.0 {
            return;
        }
        let gmax = array.effective_g_max();
        let rows = array.rows();
        let cols = array.cols();
        for i in 0..rows {
            for j in 0..cols {
                let np = self.dist.sample(&mut self.rng);
                let nn = self.dist.sample(&mut self.rng);
                array.g_pos[i as usize][j as usize] =
                    (array.g_pos[i as usize][j as usize] + np).clamp(0.0, gmax);
                array.g_neg[i as usize][j as usize] =
                    (array.g_neg[i as usize][j as usize] + nn).clamp(0.0, gmax);
            }
        }
    }
}

pub struct ReadDisturbSimulator {
    cfg: Config,
}

impl ReadDisturbSimulator {
    pub fn new(cfg: Config) -> Self {
        Self { cfg }
    }

    pub fn apply_disturb(&self, array: &mut CrossbarArray, active_row: i32, v_applied: f32) {
        if active_row < 0 || active_row >= array.rows() {
            return;
        }
        let v_dis = v_applied * self.cfg.disturb_ratio;
        let delta = self.cfg.disturb_alpha * v_dis;
        let gmax = array.effective_g_max();
        for neighbor in [active_row - 1, active_row + 1] {
            if neighbor < 0 || neighbor >= array.rows() {
                continue;
            }
            for j in 0..array.cols() {
                let ni = neighbor as usize;
                let nj = j as usize;
                array.g_pos[ni][nj] = (array.g_pos[ni][nj] + delta).clamp(0.0, gmax);
                array.g_neg[ni][nj] = (array.g_neg[ni][nj] + delta).clamp(0.0, gmax);
            }
        }
    }

    pub fn log_drift_report(&self, array: &CrossbarArray) {
        let n = (array.rows() * array.cols()) as i32;
        if n == 0 {
            println!("[ReadDisturb] drift report: empty array");
            return;
        }
        let mut sum_shift = 0.0f64;
        let mut max_shift = 0.0f64;
        for i in 0..array.rows() {
            for j in 0..array.cols() {
                let gp0 = array.g_pos_baseline[i as usize][j as usize];
                let gn0 = array.g_neg_baseline[i as usize][j as usize];
                let gp = array.g_pos[i as usize][j as usize];
                let gn = array.g_neg[i as usize][j as usize];
                let shift_pos = (gp as f64 - gp0 as f64).abs();
                let shift_neg = (gn as f64 - gn0 as f64).abs();
                let cell_shift = shift_pos.max(shift_neg);
                sum_shift += cell_shift;
                max_shift = max_shift.max(cell_shift);
            }
        }
        let avg = sum_shift / n as f64;
        println!(
            "[ReadDisturb] drift: avg conductance shift = {avg} S, max = {max_shift} S"
        );
    }
}

pub struct WriteEnduranceSimulator {
    cfg: Config,
}

impl WriteEnduranceSimulator {
    pub fn new(cfg: Config) -> Self {
        Self { cfg }
    }

    pub fn apply_write_cycles(&self, array: &mut CrossbarArray, cycles: i32) {
        if cycles <= 0 || self.cfg.write_endurance_lambda <= 0.0 {
            return;
        }
        let scale = (-self.cfg.write_endurance_lambda * cycles as f32).exp();
        array.apply_uniform_conductance_scale(scale);
    }
}
